Prefer pods without a deletion timestamp over pods with a deletion timestamp when getting an ordered list of active pods

brianpursley opened the pull request:
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
Prefer pods without a deletion timestamp over pods with a deletion timestamp when getting an ordered list of active pods.

This allows commands like kubectl port-forward, when used with deployments, to avoid selecting a pod that is terminating (due to a rollout, for example).

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubectl/issues/1409

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Fixed a bug where `kubectl port-forward`, when used with a Deployment, could connect to a terminating pod even when a running pod is also available.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
None

brianpursley commented:
/hold for review

ardaguclu commented:
Nice catch!
LGTM

brianpursley commented:
/retest

ardaguclu commented:
/lgtm

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go:
```diff
@@ -142,18 +146,26 @@ func (s ActivePods) Less(i, j int) bool {
+		return isPodDeleting(s[i])
+	}
+	// 5. Older deletion timestamp < newer deletion timestamp
+	if isPodDeleting(s[i]) && isPodDeleting(s[j]) && !s[i].ObjectMeta.DeletionTimestamp.Equal(s[j].ObjectMeta.DeletionTimestamp) {
```
I think we can remove `!s[i].ObjectMeta.DeletionTimestamp.Equal(s[j].ObjectMeta.DeletionTimestamp)` condition for the simplicity because it is very rare that `s[i].ObjectMeta.DeletionTimestamp` and `s[j].ObjectMeta.DeletionTimestamp`  are equal and even this happens, we can continue to other conditions enumerated below.

brianpursley replied:
Thanks for that comment.

The reason I original added that condition was to follow the pattern set by the comparison of ready time below (step 6).

At first, I agreed with your comment and updated this PR, but then I updated it back because I'm concerned that if pod deletion timestamp is not nil and also equal, it won't fall through to the other conditions then, and it will arbitrarily choose one of the deleting pods without any consideration for the other factors.

Removing that condition could make the active pod comparisons "unstable", in a stable-sort sort of way... By that, I mean calling it multiple times with the same elements in a different starting order could yield a different result, when ideally it should be the same result regardless of the input order.

I agree that it seems unlikely for two pods to have **exactly** the same deletion timestamp, but it could happen with an automated process that bulk-deletes pods for example.

Anyway, I think it seems safer to leave this condition in.  I think it is worth the reduced simplicity to be explicit about how this case is handled.

My two pushes (change and change back) removed your lgtm, so if you still think this is OK, another lgtm would be appreciated! :sweat_smile:

ardaguclu replied:
It makes sense and I agree with that(especially comparing with the step 6 which applies similar rule);
/lgtm

brianpursley committed:
Prefer pods without a deletion timestamp over pods with a deletion timestamp when getting an ordered list of active pods.

This allows commands like kubectl port-forward, when used with deployments, to avoid selecting a pod that is terminating (due to a rollout, for example).

brianpursley commented:
/unhold
/priority important-longterm

sftim commented:
Changelog suggestion
```markdown
Fixed a bug where `kubectl port-forward`, when used with a Deployment, could connect to a terminating
pod even when a running pod is also available.
```

brianpursley commented:
Good suggestion.  I updated the wording.  Thanks, @sftim!
